### PR TITLE
Block Users without Permission from Creating Environment Locks

### PR DESCRIPTION
**Pull request by gsandok**
Takes the extracted role information from the GRPC context. 
When Dex is enabled then checks if associated role has permissions for createEnvironmentLock. 
Blocks user if their role does not have permissions.

### CODE REVIEW

File: services/cd-service/pkg/service/batch.go [hunk 1]
```diff
@@ -93,11 +93,17 @@ func ValidateApplication(
 }
 
 func (d *BatchServer) processAction(
-	batchAction *api.BatchAction,
+	batchAction *api.BatchAction, user *auth.User,
 ) (repository.Transformer, *api.BatchResult, error) {
 	switch action := batchAction.Action.(type) {
 	case *api.BatchAction_CreateEnvironmentLock:
 		act := action.CreateEnvironmentLock
+		if d.RBACConfig.DexEnabled {
```
Comment (hannesg) [inline services/cd-service/pkg/service/batch.go:101]: I would avoid inlining this whole check and maybe just have an iterface that you call to authorize this call.

File: services/cd-service/pkg/service/batch.go [hunk 2]
```diff
@@ -93,11 +93,17 @@ func ValidateApplication(
...
+			_, permissionsExist := d.RBACConfig.Policy[fmt.Sprintf("p,%s,EnvironmentLock,Create,%s,allow", user.DexAuthContext.Role, act.Environment)]
+			if !permissionsExist {
+				return nil, nil, fmt.Errorf("user does not have permissions to create an environment lock")
```
Comment (hannesg) [inline services/cd-service/pkg/service/batch.go:104]: We definitely a dedicated error type for permission errors.

**Review by hannesg — COMMENTED** [review]
Please add some tests.

File: services/cd-service/pkg/service/batch_test.go [hunk 3]
```diff
@@ -145,73 +203,73 @@ func TestBatchServiceWorks(t *testing.T) {
...
-				if exists {
-					t.Error("lock was not deleted")
-				}
-			}
-			// check that the applock was created/deleted
-			{
-				appLocks, err := svc.Repository.State().GetEnvironmentApplicationLocks("production", "test")
-				if err != nil {
-					t.Fatal(err)
+			if err == nil {
```
Comment (hannesg) [inline services/cd-service/pkg/service/batch_test.go:221]: You can avoid indenting here by just returning from that function in the line before if there is an error.

File: pkg/auth/rbac.go [hunk 4]
```diff
@@ -27,8 +27,8 @@ import (
 type RBACConfig struct {
 	// Indicates if Dex is enabled.
 	DexEnabled bool
-	// The RBAC policy. In later stories mapping for policies will be integrated here
-	// Policy map[string]*Permission
+	// The RBAC policy.
```
Comment (sven-urbanski-freiheit-com) [inline pkg/auth/rbac.go:30]:
```suggestion
	// The RBAC policy. Key is for example "p,Developer,EnvironmentLock,Create,production,allow"
```

File: services/cd-service/pkg/service/batch_test.go [hunk 5]
```diff
@@ -26,10 +26,12 @@ import (
 
 	"github.com/freiheit-com/kuberpult/services/cd-service/pkg/testutil"
 	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc/codes"
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/batch_test.go:29]: overall this PR is fine, but it's really hard to test.
I discussed with @gsandok an approach to solve this.
Basically we will provide a way mock the authentication handler (dex), and overwrite with a role.
This way we can test this locally without having to setup dex.
(And my hidden agenda here is that this may be useful for technical users too).

File: pkg/auth/auth.go [hunk 6]
```diff
@@ -82,10 +82,31 @@ func WriteUserRoleToGrpcContext(ctx context.Context, userRole string) context.Co
...
+
+type DexGrpcContextReader struct {
+}
+
+type DummyGrpcContextReader struct {
+}
+
+func (x *DummyGrpcContextReader) ReadUserFromGrpcContext(ctx context.Context, dexEnabled bool) (*User, error) {
+	user := &User{
+		Email: "userMail",
```
Comment (sven-urbanski-freiheit-com) [inline pkg/auth/auth.go:97]:
```suggestion
		Email: "dummyMail@example.com",
```

File: services/cd-service/pkg/cmd/server.go [hunk 7]
```diff
@@ -125,8 +126,17 @@ func RunServer() {
 			logger.FromContext(ctx).Fatal("azure.auth.error: pgpKeyRing is required to authenticate manifests when \"KUBERPULT_AZURE_ENABLE_AUTH\" is true")
 		}
 
+		var reader auth.GrpcContextReader
+		if c.DexMock {
+			c.DexEnabled = true
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/cmd/server.go:131]: rather than changing a configuration silently, please return an error if they configs do not match:
```suggestion
		if c.DexMock {
		        if !c.DexEnabled {
		            return nil, errors.New("dexEnabled must be true if dexMock is true")
		        }
```

File: services/cd-service/pkg/cmd/server.go [hunk 8]
```diff
@@ -125,8 +126,17 @@ func RunServer() {
...
+			reader = &auth.DummyGrpcContextReader{}
+		} else {
+			reader = &auth.DexGrpcContextReader{}
+		}
+		dexRbacPolicy := map[string]*auth.Permission{}
 		if c.DexEnabled {
-			dexRbacPolicy, err := c.readRbacPolicy()
+			logger.FromContext(ctx).Info("IT WORKED :)")
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/cmd/server.go:138]: remove logging please

File: services/cd-service/pkg/service/batch.go [hunk 9]
```diff
@@ -92,12 +93,37 @@ func ValidateApplication(
 	return nil
 }
 
+func (d *BatchServer) checkUserPermissions(user *auth.User, act *api.CreateEnvironmentLockRequest, application, action string) error {
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/batch.go:96]:
```suggestion
func (d *BatchServer) checkUserPermissions(user *auth.User, env string, application, action string) error {
```
You're only using the environment.

File: services/cd-service/pkg/service/batch.go [hunk 10]
```diff
@@ -92,12 +93,37 @@ func ValidateApplication(
...
+		return err
+	}
+	group := act.Environment
+	temp := groups[act.Environment].EnvironmentGroup
+	if temp != nil {
+		group = *temp
+	}
+	_, permissionsExist := d.RBACConfig.Policy[fmt.Sprintf("p,%s,%s,%s,%s,allow", user.DexAuthContext.Role, application, action, fmt.Sprintf("%s:%s", act.Environment, group))]
+	if !permissionsExist {
+		return status.Errorf(codes.PermissionDenied, "user does not have permissions to create an environment lock")
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/batch.go:111]: please return more details here, like which line you were looking for in the rbacconfig.Policy.

File: services/cd-service/pkg/service/batch_test.go [hunk 11]
```diff
@@ -145,29 +205,30 @@ func TestBatchServiceWorks(t *testing.T) {
...
+			tc.svc.Repository = repo
+			resp, err := tc.svc.ProcessBatch(
+				tc.context,
 				&api.BatchRequest{
 					Actions: tc.Batch,
 				},
 			)
+			if err != nil && err.Error() == tc.expectedError {
+				return
+			}
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/batch_test.go:221]: Please invert this and do not just "return"
```suggestion
			if err != nil && err.Error() != tc.expectedError {
				t.Fatal(err.Error())
			}
```

**Review by hannesg — COMMENTED** [review]
I need to rethink this testing approach a bit. Maybe it's easier to not pass this reader around but instead have an undocumented flag on the auth package that enables mocking? At least, we definitely don't need to pass this reader around twice. The interceptor should put the verified user in the context and then it shouldn't matter how it got there.

File: services/cd-service/pkg/service/batch.go (hunk 9 above)
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/service/batch.go:96]: Ideally, this function should be part of the `GrpcContextReader`.
But we can do that in another PR.

File: services/cd-service/pkg/service/batch.go [hunk 12]
```diff
@@ -92,12 +93,37 @@ func ValidateApplication(
...
+	groups, err := d.Repository.State().GetEnvironmentConfigs()
+	if err != nil {
+		return err
+	}
+	group := act.Environment
+	temp := groups[act.Environment].EnvironmentGroup
+	if temp != nil {
+		group = *temp
+	}
+	_, permissionsExist := d.RBACConfig.Policy[fmt.Sprintf("p,%s,%s,%s,%s,allow", user.DexAuthContext.Role, application, action, fmt.Sprintf("%s:%s", act.Environment, group))]
```
[the remark this answers is not in the document]
Comment (gsandok) [reply services/cd-service/pkg/service/batch.go]: for which aspect of this are you asking about the format?

[the remark this answers is not in the document]
Comment (gsandok) [reply services/cd-service/pkg/service/batch.go]: making adjustment know to handle case

[the remark this answers is not in the document]
Comment (hannesg) [reply services/cd-service/pkg/service/batch.go]: Bernardo

File: services/cd-service/pkg/service/batch.go [hunk 13]
```diff
@@ -229,7 +255,7 @@ func (d *BatchServer) ProcessBatch(
 	ctx context.Context,
 	in *api.BatchRequest,
 ) (*api.BatchResponse, error) {
-	user, err := auth.ReadUserFromGrpcContext(ctx, d.RBACConfig.DexEnabled)
+	user, err := d.Reader.ReadUserFromGrpcContext(ctx, d.RBACConfig.DexEnabled)
```
[the remark this answers is not in the document]
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/service/batch.go:265]: Yes I think you're right.
The interceptor already does put it in the context. So we can here just read from context (not grpc-context) @gsandok

File: services/cd-service/pkg/service/batch.go (hunk 9 above)
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/service/batch.go]: we'll do this later

File: services/cd-service/pkg/service/batch.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/service/batch.go]: @hannesg What does that mean?
It works for now, so we'll keep it as it is.

File: services/cd-service/pkg/service/batch.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/service/batch.go:265]: done

**Comment by sven-urbanski-freiheit-com** [comment]
> At least, we definitely don't need to pass this reader around twice.

We don't do that anymore.
For the other part of your comment: We'll address this in another PR.

**Review by hannesg — APPROVED** [review]

**Review by sven-urbanski-freiheit-com — APPROVED** [review]
